### PR TITLE
[SPARK-24058][ML][PySpark] Default Params in ML should be saved separately: Python API

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

See SPARK-23455 for reference. Now default params in ML are saved separately in metadata file in Scala. We must change it for Python for Spark 2.4.0 as well in order to keep them in sync.

## How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @jkbradley @dbtsai

File: python/pyspark/ml/util.py [hunk 1]
```diff
@@ -417,15 +419,24 @@ def _get_metadata_to_save(instance, sc, extraMetadata=None, paramMap=None):
...
-        params = instance.extractParamMap()
+
+        # User-supplied param values
+        params = instance._paramMap
         jsonParams = {}
         if paramMap is not None:
             jsonParams = paramMap
         else:
             for p in params:
                 jsonParams[p.name] = params[p]
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: I think use `_paramMap.copy()` will be simpler.

File: python/pyspark/ml/util.py [hunk 2]
```diff
@@ -417,15 +419,24 @@ def _get_metadata_to_save(instance, sc, extraMetadata=None, paramMap=None):
...
+
+        # Default param values
+        jsonDefaultParams = {}
+        for p in instance._defaultParamMap:
+            jsonDefaultParams[p.name] = instance._defaultParamMap[p]
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: similar, use `_defaultParamMap.copy()`

File: python/pyspark/ml/util.py [hunk 3]
```diff
@@ -523,11 +534,29 @@ def getAndSetParams(instance, metadata):
...
+        minor = majorAndMinorVersions[1]
+        # For metadata file prior to Spark 2.4, there is no default section.
+        if major > 2 or (major == 2 and minor >= 4):
+            assert 'defaultParamMap' in metadata, "Error loading metadata: Expected " + \
+                "`defaultParamMap` section not found"
+
+            for paramName in metadata['defaultParamMap']:
+                param = instance.getParam(paramName)
+                paramValue = metadata['defaultParamMap'][paramName]
+                instance._setDefault(**{param.name: paramValue})
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: remove line `param = instance.getParam(paramName)` and change this line to
`instance._setDefault(**{paramName: paramValue})`

Comment (viirya) [reply python/pyspark/ml/util.py]: Oh, right. My bad.

File: python/pyspark/ml/util.py (hunk 1 above)
Comment (viirya) [reply python/pyspark/ml/util.py]: `_paramMap`'s keys are `Param` not string.

File: python/pyspark/ml/util.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/ml/util.py]: ditto.

File: python/pyspark/util.py [hunk 4]
```diff
@@ -61,6 +62,26 @@ def _get_argspec(f):
     return argspec
 
 
+def majorMinorVersion(version):
```
Comment (jkbradley) [inline python/pyspark/util.py]: Since this affects Spark core, it might be nice to put this in a separate PR.  Also, shall we make this match the Scala API?

**Review by jkbradley — COMMENTED** [review]
I added a quick comment about moving version utils to a separate PR.  I'll try to review the other parts soon.  Thanks!

File: python/pyspark/util.py (hunk 4 above)
Comment (viirya) [reply python/pyspark/util.py]: > Also, shall we make this match the Scala API?
Do you mean to throw exception when unable to parse Spark version?

**Comment by viirya** [comment]
Done. Can you try again?

**Comment by viirya** [comment]
Not sure why AppVeyor build failed. How to re-trigger AppVeyor build?

**Review by jkbradley — COMMENTED** [review]
Sorry for the delay.  This LGTM except for the small typo  (Hopefully fixing it will re-run the Appveyor test)

File: python/pyspark/ml/util.py [hunk 5]
```diff
@@ -396,6 +397,7 @@ def saveMetadata(instance, path, sc, extraMetadata=None, paramMap=None):
         - sparkVersion
         - uid
         - paramMap
+        - defalutParamMap (since 2.4.0)
```
[the remark this answers is not in the document]
Comment (viirya) [reply python/pyspark/ml/util.py]: Thanks.

**Comment by viirya** [comment]
OK. Looks like AppVeyor build is ok now.

**Comment by jkbradley** [comment]
OK thanks @viirya !
Merging with master

**Comment by viirya** [comment]
Thanks @jkbradley @WeichenXu123
